### PR TITLE
Add moodle webview

### PR DESCRIPTION
**Pull request by kurwjan**
Adding a WebView for Moodle (aka a minimal restricted browser?) instead of redirecting to the login website.

We need to use the `flutter_inappwebview` package because Lanis is buggy with the official WebView package and we have more options. 

iOS also shouldn't be a hassle, there are more infos in the [docs](https://inappwebview.dev/docs/intro#setup-ios). (normally it should work without editing configs)

Issues:
- ~Currently doesn't use Material 3 Components, instead I get M2 components although I specified M3.~ Uses a mix of M2 and M3, sometimes weird colour choices but it isn't an issue.
- Need to switch `flutter_inappwebview` with `webview_inapp` temporarily because it doesn't build in release mode.

Todos:
- [x] Use Material Compontents for dialogs, checkboxes, yada yada... because the default components come straight out of 2011 with the old Holo theme (iOS shouldn't have this problem)
- [x] ~More sophisticated URL launching?~ enough
- [x] Prevent logout because it will also destroy the session of the App. (probably need to use JS injection)
- [x] Back, forward and refresh button
- [x] Loading indicators
- [x] ~[Login redirect](https://moodledev.io/general/app/development/link-handling/deep-linking) to the official Moodle app ([Github](https://github.com/moodlehq/moodleapp), [Play Store](https://play.google.com/store/apps/details?id=com.moodle.moodlemobile))~
   - ~for iOS the URL scheme `moodlemobile` must be allowed, more info [here](https://pub.dev/packages/url_launcher#configuration)~

### CODE REVIEW

File: app/lib/view/moodle.dart [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+  @override
+  void initState() {
+    super.initState();
+
+    String sessionToken = "";
+    client.jar
+        .loadForRequest(Uri.parse("https://start.schulportal.hessen.de"))
+        .then((cookies) {
+      sessionToken = cookies[1].value;
```
Comment (alessioC42) [inline app/lib/view/moodle.dart:32]: Is a hard coded index of 1 unsafe here?

File: app/lib/view/moodle.dart [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    cookieManager.setCookie(
+        url: WebUri("https://schulportal.hessen.de"),
+        name: "sid",
+        value: sessionToken,
+        isSecure: true,
+        domain: ".hessen.de");
+    cookieManager.setCookie(
+        url: WebUri("https://schulportal.hessen.de"),
+        name: "i",
+        value: "6091",
```
Comment (alessioC42) [inline app/lib/view/moodle.dart:53]: ^^ I guess this should be client.schoolID ?

**Review by alessioC42 — DISMISSED** [review]
Just an overview. Havent tested it jet

File: app/lib/view/moodle.dart (hunk 1 above)
Comment (kurwjan) [reply app/lib/view/moodle.dart:32]: should be safe (at least on my android chrome and desktop firefox) but I will code a safer option

**Comment by kurwjan** [comment]
I think I used too many ValueNotifiers. I could replace some of them with setState probably.

**Review by alessioC42 — CHANGES_REQUESTED** [review]
I am currently experiencing 2 problems:
1: When authenticating for the first time, I am unable to log in to Moodle.
2: (more optional) To better mimic the behaviour of a real browser, it would be nice to use a PopScope to trigger a back event. 

Everything else works fine.

**Comment by kurwjan** [comment]
> I am currently experiencing 2 problems: 1: When authenticating for the first time, I am unable to log in to Moodle. 2: (more optional) To better mimic the behaviour of a real browser, it would be nice to use a PopScope to trigger a back event.
> 
> Everything else works fine.

Fixed it and I think it's now ready to merge.

**Review by alessioC42 — CHANGES_REQUESTED** [review]
Das Herunterladen von Dokumenten (getestet nur mit PDFs) scheint auch nicht zu funktionieren. Diese müssten irgendwie gespeichert werden (evtl mit dem bereits existierenden handler aber dafür im app cache?) und dann mit dem Systemviewer geöffnet werden

**Comment by kurwjan** [comment]
interessant, aber ich kann aber versichern das Hochladen funktioniert.

**Comment by kurwjan** [comment]
Das war eine Achterbahn und am Ende war es eigentlich so einfach gewesen:
```dart
final moodleCookie1 = await CookieManager.instance().getCookie(url: WebUri("https://mo${client.schoolID}.schulportal.hessen.de"), name: "MoodleSession");
final dioCookie1 = dio.Cookie(moodleCookie1!.name, moodleCookie1.value);
dioCookie1.httpOnly = false;
dioCookie1.secure = true;
dioCookie1.domain = "mo${client.schoolID}.schulportal.hessen.de";
dioCookie1.path = "/";

final moodleCookie2 = await CookieManager.instance().getCookie(url: WebUri("https://mo${client.schoolID}.schulportal.hessen.de"), name: "MOODLEID1_");
final dioCookie2 = dio.Cookie(moodleCookie2!.name, moodleCookie2.value);
dioCookie2.httpOnly = false;
dioCookie2.secure = true;
dioCookie2.domain = "mo${client.schoolID}.schulportal.hessen.de";
dioCookie2.path = "/";

final moodleCookie3 = await CookieManager.instance().getCookie(url: WebUri("https://mo${client.schoolID}.schulportal.hessen.de"), name: "mo-prod01");
final dioCookie3 = dio.Cookie(moodleCookie3!.name, moodleCookie3.value);
dioCookie3.httpOnly = true;
dioCookie3.secure = true;
dioCookie3.domain = ".hessen.de";
dioCookie3.path = "/";

await client.jar.saveFromResponse(Uri.parse("https://mo${client.schoolID}.schulportal.hessen.de"), [dioCookie1, dioCookie2]);
await client.jar.saveFromResponse(Uri.parse("https://start.schulportal.hessen.de"), [dioCookie3]);
```
Dateien sollte man jetzt downloaden können

**Pull request merged**